### DISCUSSION TITLE
PS3 Quick Menu

### DISCUSSION THREAD

**Discussion by AkiraKusanagi** [EmiyaSyahriel/CrossLauncher#79] [Ideas]
It would be cool if holding the Home/PS Button (Also called Pad1_Mode on input mapping stuff) Put up a quick menu with options like the PS3.
I imagine it would need these permissions:

- App Usage Data
- Display Over Other Apps
- Modify System Settings (Possibly, but, I'm not really sure)

Features it should have:

- Quit App
- Turn Off Controller
- Settings

**Comment by EmiyaSyahriel** [comment]
Sure, But my plan is that instead of the limited menu as in PS3, there will be the full launcher displayed. Or maybe user should be able to change what to display here?

### However, there is some limitations :
And I'd like to ask how would you and possibly other users would liked it.

#### &bullet; Quit App
Forcefully quitting other apps will probably requires either one or a combination of :
  - Shizuku access
  - Root access
  - System app permission, The launcher might need to be installed as system app, either by rooting, or as a part of a custom or OEM ROM.
  - System code reflection, but this is unstable as Android or Manufacturers can change their method names at a certain version of their ROM which is crucial for reflection.
 
The "Force Close" option at the app's side menu is only "sending close request", it's up to the system to close it or not and it only applies to background services, not the one visible to user. I should've change to a better wording, or maybe I should check in those "RAM Cleaner" apps. 

If you have found any app that is able to do this without the said limitation and is open source, you can specify it in this discussion.

#### &bullet; Listening for the Button itself
So far, I can't find any method to listen any system-wide button presses, not even in the Accessibility API that is usually used in this regard and some that I've found end up having limitations as the point above. I haven't tested if a floating window can also listen for button presses if they are not active / focused though.

Again, If you have found any app that is able to do this without the said limitation and is open source, you can specify it in this discussion.

#### &bullet; Turn Off Controller
This might be only available on Bluetooth devices since most bluetooth gamepads will turn off if it is disconnected. *(still have trauma with that Bootleg PS3 Controller though)*, would that be okay?

As for wired, you cannot control the controller to the level you can turn it off in Android unless you disconnects the cable itself. Some controllers (e.g DualSense) did supports a certain command to change it's hardware state (e.g Adaptive Trigger) but not it's power state. Well some might have this command to turn theirselves off, but I still can't find one sample for it.
